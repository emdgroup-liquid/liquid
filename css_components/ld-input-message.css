.ld-input-message,:host{--ld-input-message-valid-text-col:var(--ld-thm-success);--ld-input-message-error-text-col:var(--ld-thm-error);display:inline-flex;align-items:baseline;font:var(--ld-typo-body-s);line-height:1.5}.ld-input-message--info .ld-input-message__icon,:host(.ld-input-message--info) .ld-input-message__icon{--ld-icon-secondary-col:var(--ld-col-neutral-900);color:var(--ld-thm-warning)}.ld-input-message--valid,:host(.ld-input-message--valid){color:var(--ld-input-message-valid-text-col)}.ld-input-message--error,:host(.ld-input-message--error){color:var(--ld-input-message-error-text-col)}.ld-input-message__icon{transform:translateY(var(--ld-sp-2));margin-top:var(--ld-sp-2);margin-right:var(--ld-sp-6);flex-shrink:0}