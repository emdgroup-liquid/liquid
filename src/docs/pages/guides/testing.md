---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

