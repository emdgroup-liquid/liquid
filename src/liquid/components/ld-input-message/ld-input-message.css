:host,
.ld-input-message {
  --ld-input-message-valid-text-col: var(--ld-thm-success);
  --ld-input-message-error-text-col: var(--ld-thm-error);

  display: inline-flex;
  align-items: baseline;
  font: var(--ld-typo-body-s);
  line-height: 1.5;
}

:host(.ld-input-message--valid),
.ld-input-message--valid {
  color: var(--ld-input-message-valid-text-col);
}

:host(.ld-input-message--error),
.ld-input-message--error {
  color: var(--ld-input-message-error-text-col);
}

.ld-input-message__icon {
  transform: translateY(var(--ld-sp-2));
  margin-top: var(--ld-sp-2);
  margin-right: var(--ld-sp-6);
  flex-shrink: 0;
}
