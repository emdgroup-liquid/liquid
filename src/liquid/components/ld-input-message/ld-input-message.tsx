import '../../components' // type definitions for type checks and intelliSense
import { Component, h, Prop, getAssetPath } from '@stencil/core'

@Component({
  assetsDirs: ['assets'],
  tag: 'ld-input-message',
  styleUrl: 'ld-input-message.css',
  shadow: false,
})
export class LdInputMessage {
  /** Input message mode. */
  @Prop() mode: 'error' | 'info' | 'valid' = 'error'

  /**
   * This property does **not** change the visual appearance of the input message.
   * Set this property to `true` if the message is hidden initially and change it to
   * `false` as soon as you make the message visible again. It changes the `aria-hidden`
   * attribute on the slot which is wrapped by an element that has an `aria-live`
   * attribute with the value `assertive`. This means that the slot is wrapped by a
   * so called “live region”, which is conveyed to screen readers and other assistive
   * technology. The value “assertive” emphasizes the importance of the message and
   * causes screen readers to interrupt their current tasks to read aloud this message.
   * Thus the message is read aloud before the next element that received the focus
   * is announced to the user.
   */
  @Prop() covert = false

  render() {
    let cl = 'ld-input-message'
    cl += ` ld-input-message--${this.mode}`

    return (
      <span class={cl}>
        <img
          class="ld-input-message__icon"
          alt=""
          role="presentation"
          src={getAssetPath(`./assets/${this.mode}.svg`)}
        />
        <span aria-live="assertive">
          <span aria-hidden={this.covert ? 'true' : undefined}>
            <slot></slot>
          </span>
        </span>
      </span>
    )
  }
}
