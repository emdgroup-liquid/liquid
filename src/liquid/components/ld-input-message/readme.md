---
eleventyNavigation:
  key: Input Message
  parent: Components
layout: layout.njk
title: Input Message
permalink: components/ld-input-message/
---

# ld-input-message

<link rel="stylesheet" href="{{ env.base }}/{{ buildstamp }}css_components/ld-input-message.css">
<link rel="stylesheet" href="{{ env.base }}/{{ buildstamp }}css_components/ld-icon.css">

This component is meant to be used in conjunction with the [`ld-input`](components/ld-input/) and the [`ld-label`](components/ld-label/) component. Please reffer to the [`ld-input`](components/ld-input/) docs for further usage examples.

<ld-notice headline="Note" mode="warning">
  The <code>ld-input-message</code> with a mode other than <code>"info"</code> is supposed to be used on white background only, as on other backgrounds the color contrast of the text is insufficient against the background. 
</ld-notice>

<ld-notice headline="Note" mode="warning">
  If you choose to use the CSS component and plan to conditionally show the message, such as when an input field becomes invalid after user interaction, you will need to make sure that screen readers or other assistive technology become aware of a message as soon as it becomes visible. You might want to use <a href="https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions" rel="noreferrer noopener" target="_blank">ARIA live regions</a> for this matter.
</ld-notice>

## Examples

### As error message

{% example '{ "background": "light" }' %}
<ld-input-message>This field is required.</ld-input-message>

<!-- React component -->

<LdInputMessage>This field is required.</LdInputMessage>

<!-- CSS component -->

<span class="ld-input-message ld-input-message--error">
  <!-- Note that you can use an img element with the class ld-input-message__icon here, as well. -->
  <svg class="ld-input-message__icon ld-icon ld-icon--sm" width="14" height="14" viewBox="0 0 14 14" fill="none">
    <path fill-rule="evenodd" clip-rule="evenodd" d="M7 14C10.866 14 14 10.866 14 7C14 3.13401 10.866 0 7 0C3.13401 0 0 3.13401 0 7C0 10.866 3.13401 14 7 14Z" fill="currentColor"/>
    <path d="M4.66675 4.66699L9.33341 9.33366" stroke="var(--ld-col-wht)" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"/>
    <path d="M4.66675 9.33301L9.33341 4.66634" stroke="var(--ld-col-wht)" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"/>
  </svg>
  This field is required.
</span>
{% endexample %}

### As info message

{% example '{ "background": "light" }' %}
<ld-input-message mode="info">This field will destroy itself on form submission.</ld-input-message>

<!-- React component -->

<LdInputMessage mode="info">This field will destroy itself on form submission.</LdInputMessage>

<!-- CSS component -->

<span class="ld-input-message ld-input-message--info">
  <!-- Note that you can use an img element with the class ld-input-message__icon here, as well. -->
  <svg class="ld-input-message__icon ld-icon ld-icon--sm" width="14" height="14" viewBox="0 0 14 14" fill="none">
    <path fill-rule="evenodd" clip-rule="evenodd" d="M7 14C10.866 14 14 10.866 14 7C14 3.13401 10.866 0 7 0C3.13401 0 0 3.13401 0 7C0 10.866 3.13401 14 7 14Z" fill="var(--ld-thm-warning)"/>
    <path fill-rule="evenodd" clip-rule="evenodd" d="M7.18234 11.0254C6.79228 11.0254 6.48657 10.9147 6.26518 10.6933C6.0438 10.472 5.93311 10.1662 5.93311 9.77618V6.12335C5.93311 5.99685 6.0069 5.93359 6.15449 5.93359H6.89771C7.28776 5.93359 7.59348 6.04428 7.81487 6.26567C8.03625 6.48705 8.14694 6.79277 8.14694 7.18283V10.8357C8.14694 10.9622 8.07315 11.0254 7.92556 11.0254H7.18234Z" fill="var(--ld-col-neutral-900)"/>
    <ellipse cx="6.99977" cy="3.80007" rx="1.06667" ry="1.06667" fill="var(--ld-col-neutral-900)"/>
  </svg>
  This field will destroy itself on form submission.
</span>
{% endexample %}

### As success message

{% example '{ "background": "light" }' %}
<ld-input-message mode="valid">That's correct!</ld-input-message>

<!-- React component -->

<LdInputMessage mode="valid">That's correct!</LdInputMessage>

<!-- CSS component -->

<span class="ld-input-message ld-input-message--valid">
  <!-- Note that you can use an img element with the class ld-input-message__icon here, as well. -->
  <svg class="ld-input-message__icon ld-icon ld-icon--sm" width="14" height="14" viewBox="0 0 14 14" fill="none">
    <circle cx="7" cy="7" r="7" fill="currentColor"/>
    <path stroke="var(--ld-col-wht)" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M10.1 5.1L6.1 9 3.8 6.8"/>
  </svg>
  That's correct!
</span>
{% endexample %}


<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description                                              | Type                           | Default     |
| -------- | --------- | -------------------------------------------------------- | ------------------------------ | ----------- |
| `key`    | `key`     | for tracking the node's identity when working with lists | `string \| number`             | `undefined` |
| `mode`   | `mode`    | Input message mode.                                      | `"error" \| "info" \| "valid"` | `'error'`   |
| `ref`    | `ref`     | reference to component                                   | `any`                          | `undefined` |


## Shadow Parts

| Part     | Description                 |
| -------- | --------------------------- |
| `"icon"` | Image tag used for the icon |


## Dependencies

### Used by

 - [ld-file-upload](../ld-file-upload)
 - [ld-upload-item](../ld-file-upload/ld-upload-item)

### Depends on

- [ld-icon](../ld-icon)

### Graph
```mermaid
graph TD;
  ld-input-message --> ld-icon
  ld-file-upload --> ld-input-message
  ld-upload-item --> ld-input-message
  style ld-input-message fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
