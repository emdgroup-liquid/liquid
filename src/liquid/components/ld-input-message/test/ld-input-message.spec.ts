import { newSpecPage } from "@stencil/core/testing";
import { LdInputMessage } from "../ld-input-message";

describe("ld-input-message", () => {
  it("renders as error message by default", async () => {
    const page = await newSpecPage({
      components: [LdInputMessage],
      html: `<ld-input-message>This field is required.</ld-input-message>`,
    });
    expect(page.root).toMatchSnapshot();
  });
  it("applies the mode modifier, if mode is set", async () => {
    const page = await newSpecPage({
      components: [LdInputMessage],
      html: `<ld-input-message mode="info">This field is optional.</ld-input-message>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
