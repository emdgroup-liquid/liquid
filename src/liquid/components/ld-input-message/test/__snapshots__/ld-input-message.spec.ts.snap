// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-input-message applies the mode modifier, if mode is set 1`] = `
<ld-input-message class="ld-input-message ld-input-message--info" mode="info">
  <mock:shadow-root>
    <ld-icon class="ld-input-message__icon" name="ld-input-message-info" part="icon" size="sm"></ld-icon>
    <span aria-live="assertive">
      <slot></slot>
    </span>
  </mock:shadow-root>
  This field is optional.
</ld-input-message>
`;

exports[`ld-input-message renders as error message by default 1`] = `
<ld-input-message class="ld-input-message ld-input-message--error">
  <mock:shadow-root>
    <ld-icon class="ld-input-message__icon" name="ld-input-message-error" part="icon" size="sm"></ld-icon>
    <span aria-live="assertive">
      <slot></slot>
    </span>
  </mock:shadow-root>
  This field is required.
</ld-input-message>
`;
